### PR TITLE
Major WebcamVideoStream and PiVideoStream classes - threading bugfixes, resource releasing and Other major fixes.

### LINKED ISSUE #35: WebcamVideoStream not releasing the camera when stopped
Hello,
It looks like the WebcamVideoStream is not releasing the VideoCapture in the stop method. So when I exit out of python the camera is not being released.
Changing the stop function like this fixed the issue for me:

class WebcamVideoStream:
                  ....
def stop(self):
		# indicate that the thread should be stopped
		self.stopped = True
		self.stream.release()

### PR DESCRIPTION
**Pull request by abhiTronix**
Hello @jrosebr1, Thanks again for this awesome library.

This PR fixes #35 #38 #47 and many more.

## Bug
I've been working on my project and encountered some similar **resource mishandling problems in WebcamVideoStream class and VideoStream class** too. The main problem was that the video stream and thread resources not handled properly in WebcamVideoStream class and further digging into it I noticed some code inconsistency related to it in the library resulting in the following error after every execution:
```
FATAL: exception not rethrown
Aborted (core dumped)
```
### Fixes
This problem is solved in FileVideoStream class in following PR #86, But these two class mentioned seems to remain unaffected and still throws the error. So after several tests, I came up with the following PR(_Without any mistakes like the previous one_) and fixed these issues once for all.  Kudos

### CODE REVIEW

**Comment by pageauc** [comment]
Hi Adrian

I will put a watch on this repo because I want to know when these changes are implemented.  Do you have a rough ETA
Thanks

**Comment by jrosebr1** [comment]
@abhiTronix Thanks so much for the PR. The `imutils` library is definitely a special case library as it needs to (1) provide correct functionality while (2) not breaking tutorials on PyImageSearch. Have you tried with any of the `WebcamVideoStream` posts on the PyImageSearch blog? If it breaks too much functionality with those then I cannot merge the PR.

Secondly, if this issue affects `WebcamVideoStream` then it most certainly affects `PiVideoStream` -- did you try with that class as well?

@pageauc See my comments above.

**Comment by abhiTronix** [comment]
Hello @jrosebr1, 

> (1) provide correct functionality while (2) not breaking tutorials on PyImageSearch. Have you tried with any of the WebcamVideoStream posts on the PyImageSearch blog?

Yes, **I myself already performed several tests on VideoStream & WebcamStream Class** and **it didn't cause any kind errors or exceptions**_(I'm already using a major part of your blog post for my research_). **After these tests, I'm certain from the side, about that it won't cause any harm to the existing functionality or broke the libraries bindings.** _As seen in this PR, I didn't alter the structure of existing WebcamStream Class but at the same time added extra functionality to make it more Thread safe_. **However, I will advise running the test on your side too, just to be sure.** 

> Secondly, if this issue affects WebcamVideoStream then it most certainly affects PiVideoStream -- did you try with that class as well?

No, `WebcamStream` class **not related to** `PiVideoStream` class **in any manner**(_Since both have separate threads and classes with no crossover_), However, **VideoStream Class does use both these classes for its functionality**. So looking at the `VideoStream` Class code below:
```
# import the necessary packages
from .webcamvideostream import WebcamVideoStream

class VideoStream:
	def __init__(self, src=0, usePiCamera=False, resolution=(320, 240),
		framerate=32):
		# check to see if the picamera module should be used
		if usePiCamera:
			# only import the picamera packages unless we are
			# explicity told to do so -- this helps remove the
			# requirement of `picamera[array]` from desktops or
			# laptops that still want to use the `imutils` package
			from .pivideostream import PiVideoStream

			# initialize the picamera stream and allow the camera
			# sensor to warmup
			self.stream = PiVideoStream(resolution=resolution,
				framerate=framerate)

		# otherwise, we are using OpenCV so initialize the webcam
		# stream
		else:
			self.stream = WebcamVideoStream(src=src)

	def start(self):
		# start the threaded video stream
		return self.stream.start()

	def update(self):
		# grab the next frame from the stream
		self.stream.update()

	def read(self):
		# return the current frame
		return self.stream.read()

	def stop(self):
		# stop the thread and release any resources
self.stream.stop()
```
As seen from the code, the `usePiCamera` flag in `if-else` statement completely seperates `WebcamStream` class from `PiVideoStream` class and thus they don't touch each other's functionality. **Since only either of them(_WebcamStream or PiVideoStream_) will be activated at a given moment.** _Therefore, **this PR has no  whatsoever effect on `PiVideoStream`** class._

So, in conclusion, this PR is good to go from my side :+1: , But as I mentioned above,  I will advise running the test on your side too, just to be on the safer side. Thank you

**Comment by abhiTronix** [comment]
Hello @jrosebr1, Thank you for the valuable reply

> Basically, I'm asking if the threading issue you fixed is also present in the PiVideoStream class? You improved the ability of WebcamVideoStream to handle threading but keep in mind that PiVideoStream also has a similar threading implementation. I'm wondering if PiVideoStream may suffer from the same threading problem.

Yes, I agree with you `PiVideoStream` class may have similar Threading & Resource release problem as `WebcamVideoStream` class and the code can be improved. I'd love to push another PR with improvements, But I never used `Picamera` module before nor have any access to this module right now, and therefore I can't test the results.

But **I definitely found this valuable comment by the `Picamera` library owner**: https://github.com/waveform80/picamera/issues/461#issuecomment-425578406 supporting `capture_continuous` isn't a good choice for threading?

Also, some issues related to this you may want to look at:
https://github.com/waveform80/picamera/issues?utf8=%E2%9C%93&q=is%3Aissue+capture_continuous+

I think we need seperate PR addressing & fixing `PiVideoStream` class issues in future, **What're your thoughts on this**?

**Comment by abhiTronix** [comment]
Hello @jrosebr1, 

**I have fixed several Core Bugs from `PiVideoStream` class and its threading implementations with above commits. Thereby now both `WebcamVideoStream` and `PiVideoStream` classes are now completely fixed with this PR.**

Recent commits Fixes:
#99, #111, #90  
 
I have tested these recent commits with latest PiCamera Library with my RaspiCam. Also, I've stress tested these commits with your blog post examples as well as with my own Humanoid Eyes algorithms. Plus I added the `**option flag` to enable extra features such as *Brightness, contrast control* etc and a `try-catch` block to catch any error while retrieving frames. Kindly free to merge this important PR whenever ready. 

Regards 
**_AbhiTronix_**
